src/go/k8s/pkg/resources/external_dns.go

```diff
@@ -0,0 +1,193 @@
+// Copyright 2021 Vectorized, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+// Package resources contains reconciliation logic for redpanda.vectorized.io CRD
+package resources
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	redpandav1alpha1 "github.com/vectorizedio/redpanda/src/go/k8s/apis/redpanda/v1alpha1"
+	"github.com/vectorizedio/redpanda/src/go/k8s/pkg/labels"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/scheme"
+	"sigs.k8s.io/external-dns/endpoint"
+)
+
+var (
+	// GroupVersion is group version used to register these objects
+	GroupVersion	= schema.GroupVersion{Group: "externaldns.vectorized.io", Version: "v1alpha1"}
+
+	// SchemeBuilder is used to add go types to the GroupVersionKind scheme
+	SchemeBuilder	= &scheme.Builder{GroupVersion: GroupVersion}
+
+	// AddToScheme adds the types in this group-version to the given scheme.
+	AddToScheme	= SchemeBuilder.AddToScheme
+)
+
+func init() {
+	SchemeBuilder.Register(&endpoint.DNSEndpoint{}, &endpoint.DNSEndpointList{})
+}
+
+var _ Resource = &externalDNS{}
+
+// externalDNS is part of the reconciliation of redpanda.vectorized.io CRD
+// focusing on the external connectivity management of redpanda cluster
+type externalDNS struct {
+	k8sclient.Client
+	scheme		*runtime.Scheme
+	pandaCluster	*redpandav1alpha1.Cluster
+	logger		logr.Logger
+
+	endpoints	[]*endpoint.Endpoint
+}
+
+// NewExternalDNS creates externalDNS
+func NewExternalDNS(
+	client k8sclient.Client,
+	pandaCluster *redpandav1alpha1.Cluster,
+	crlScheme *runtime.Scheme,
+	logger logr.Logger,
+) Resource {
+	return &externalDNS{
+		client,
+		crlScheme,
+		pandaCluster,
+		logger.WithValues("Kind", dnsEndpoint()),
+		nil,
+	}
+}
+
+// Ensure will manage external dns endpoint.DNSEndpoint for redpanda.vectorized.io custom resource
+func (r *externalDNS) Ensure(ctx context.Context) error {
+	if !r.pandaCluster.Spec.ExternalConnectivity {
+		return nil
+	}
+
+	var observedPods corev1.PodList
+
+	err := r.List(ctx, &observedPods, &k8sclient.ListOptions{
+		LabelSelector:	labels.ForCluster(r.pandaCluster).AsClientSelector(),
+		Namespace:	r.pandaCluster.Namespace,
+	})
+	if err != nil {
+		return fmt.Errorf("unable to list pods: %w", err)
+	}
+
+	var node corev1.Node
```

Comment:
Because it would create unnecessary node objects. Each iteration only need reference to node in order to retrieve public IP. I'm trying to not create any pressure on golang garbage collection.